Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -697,6 +697,217 @@ Java.perform(function () {
+
+When you start the App on an emulator or a rooted device, you find that the app presents a dialog box and exits as soon as you press "Ok" because it detected root: 
+
+![Crackme Root Detected Dialog](Images/Chapters/0x05c/crackme-frida-1.png)
+
+Let us see how we can prevent this.
+The decompiled main method (using CFR decompiler) looks like this:
+
+```
+package sg.vantagepoint.uncrackable1;
+
+import android.app.Activity;
+import android.app.AlertDialog;
+import android.content.Context;
+import android.content.DialogInterface;
+import android.os.Bundle;
+import android.text.Editable;
+import android.view.View;
+import android.widget.EditText;
+import sg.vantagepoint.uncrackable1.a;
+import sg.vantagepoint.uncrackable1.b;
+import sg.vantagepoint.uncrackable1.c;
+
+public class MainActivity
+extends Activity {
+    private void a(String string) {
+        AlertDialog alertDialog = new AlertDialog.Builder((Context)this).create();
+        alertDialog.setTitle((CharSequence)string);
+        alertDialog.setMessage((CharSequence)"This in unacceptable. The app is now going to exit.");
+        alertDialog.setButton(-3, (CharSequence)"OK", (DialogInterface.OnClickListener)new b(this));
+        alertDialog.show();
+    }
+
+    protected void onCreate(Bundle bundle) {
+        if (sg.vantagepoint.a.c.a() || sg.vantagepoint.a.c.b() || sg.vantagepoint.a.c.c()) {
+            this.a("Root detected!"); //This is the message we are looking for
+        }
+        if (sg.vantagepoint.a.b.a((Context)this.getApplicationContext())) {
+            this.a("App is debuggable!");
+        }
+        super.onCreate(bundle);
+        this.setContentView(2130903040);
+    }
+
+    public void verify(View object) {
+        object = ((EditText)this.findViewById(2131230720)).getText().toString();
+        AlertDialog alertDialog = new AlertDialog.Builder((Context)this).create();
+        if (a.a((String)object)) {
+            alertDialog.setTitle((CharSequence)"Success!");
+            alertDialog.setMessage((CharSequence)"This is the correct secret.");
+        } else {
+            alertDialog.setTitle((CharSequence)"Nope...");
+            alertDialog.setMessage((CharSequence)"That's not it. Try again.");
+        }
+        alertDialog.setButton(-3, (CharSequence)"OK", (DialogInterface.OnClickListener)new c(this));
+        alertDialog.show();
+    }
+}
+```
+
+Notice the `Root detetected` message in the `onCreate` method and the various methods called in the the `if`-statement before which perform the actual root checks. Also note the `This is unacceptable...` message from the first method of the class, `private void a`. Obviously, this is where the dialog box gets displayed. There is a `alertDialog.onClickListener`callback set in the `setButton` method call which is responsible for closing the application via `System.exit(0)` after successful root detection. Using Frida, we can prevent the app from exiting by hooking the callback.
```

Comment:
I don't see line numbers on other codelistings here, won't add them manually because that doesn't make sense. Line numbers were great if could add them globally.